### PR TITLE
Ywyy9

### PR DESCRIPTION
**Pull request by yinwang-wish**
- fix update command to support replace without update operator
- fix check non-required field
- update unit tests
- adding more log

### CODE REVIEW

File: pkg/mongoproxy/plugins/insort/process.go [hunk 1]
```diff
@@ -56,7 +56,7 @@ func genericLess(a, b interface{}) bool {
 		case primitive.Regex:
 			otherVal = otherValTyped.Pattern
 		default:
-			panic("what")
+			panic("cannot find correct type in Process #{valTyped}")
```
Comment (zengmin-wish) [inline pkg/mongoproxy/plugins/insort/process.go]: are you sure this works `#{valTyped}` ?, never saw such format before

File: pkg/mongoproxy/plugins/schema/type_test.go [hunk 2]
```diff
@@ -25,13 +25,13 @@ var (
...
 		// Check string array
-		{DB: "testdb", Collection: "testcollection", In: bson.D{{"friends", bson.A{"bob", "alice"}}}, Err: false},
+		{DB: "testdb", Collection: "testcollection", In: bson.D{{"friends", bson.A{"bob", "alice"}}, {"name", "tom"}}, Err: false},
 		// Check string array with wrong type
 		{DB: "testdb", Collection: "testcollection", In: bson.D{{"friends", bson.A{"bob", 1}}}, Err: true},
 		// Check int array
 		{DB: "testdb", Collection: "nonrequire", In: bson.D{{"luckynumbers", bson.A{666, 888}}}, Err: false},
 		// Check int array with wrong type
-		{DB: "testdb", Collection: "nonrequire", In: bson.D{{"luckynumbers", bson.A{666, "888"}}}, Err: true},
+		{DB: "testdb", Collection: "nonrequire", In: bson.D{{"luckynumbers", bson.A{666, "888"}}}, Err: false},
```
Comment (zengmin-wish) [inline pkg/mongoproxy/plugins/schema/type_test.go]: why this change?

File: pkg/mongoproxy/plugins/schema/types.go [hunk 3]
```diff
@@ -234,35 +234,59 @@ func (c *Collection) ValidateUpdate(ctx context.Context, obj bson.D, upsert bool
...
-				for _, item := range e.Value.(bson.D) {
-					setFields[item.Key] = item.Value
+	logrus.Debugf("pending validation object: %s", obj)
+	/*
+		update command support ,
+		* A document that contains update operator expressions,
+		* A replacement document with only <field1>: <value1> pairs
+		https://www.mongodb.com/docs/v4.2/reference/command/update/#update-statement-documents
+	*/
+	op_map := BuildUpdateOpSet()
```
Comment (zengmin-wish) [inline pkg/mongoproxy/plugins/schema/types.go]: this should be built only once, not for every check.

File: pkg/mongoproxy/plugins/schema/types.go [hunk 4]
```diff
@@ -234,35 +234,59 @@ func (c *Collection) ValidateUpdate(ctx context.Context, obj bson.D, upsert bool
...
+	if !strings.HasPrefix(obj[0].Key, "$") || !SetContain(op_map, obj[0].Key) {
```
Comment (zengmin-wish) [inline pkg/mongoproxy/plugins/schema/types.go]: isn't every key in op_map start with '$'?  what's the point of the prefix check here?

File: pkg/mongoproxy/plugins/schema/types.go [hunk 5]
```diff
@@ -234,35 +234,59 @@ func (c *Collection) ValidateUpdate(ctx context.Context, obj bson.D, upsert bool
...
+		m := make(bson.M, len(obj))
+		if upsert {
+			logrus.Debugf("upsert true, pending check")
+			insertFields = handleObj(obj, m)
+			logrus.Debugf("insertFields updated")
```
Comment (zengmin-wish) [inline pkg/mongoproxy/plugins/schema/types.go]: this looks redundant even as debug statement, it hinders readability a bit.

File: pkg/mongoproxy/plugins/schema/types.go [hunk 6]
```diff
@@ -234,35 +234,59 @@ func (c *Collection) ValidateUpdate(ctx context.Context, obj bson.D, upsert bool
...
+			logrus.Debugf("insertFields: %s", insertFields)
+		} else {
+			logrus.Debugf("upsert false with nil setFeild, pending check")
+			setFields = handleObj(obj, m)
+			logrus.Debugf("setFields updated")
```
Comment (zengmin-wish) [inline pkg/mongoproxy/plugins/schema/types.go]: same here.

File: pkg/mongoproxy/plugins/schema/types.go (hunk 5 above)
Comment (yinwang-wish) [reply pkg/mongoproxy/plugins/schema/types.go]: Add them here since, if there is any type conversion/assersion error, system will close connection directly, will have no clue where is the RC. I can remove some of them.

File: pkg/mongoproxy/plugins/schema/types.go (hunk 4 above)
Comment (yinwang-wish) [reply pkg/mongoproxy/plugins/schema/types.go]: obj[0].Key if not from op_map, it is from obj. We need to check whether the first element in the obj start with "$" or not. If not, then that means it's a replace operation. Else if it starts with "$" then need to make sure it is just a field name instead of a update operator.

File: pkg/mongoproxy/plugins/schema/types.go (hunk 3 above)
Comment (yinwang-wish) [reply pkg/mongoproxy/plugins/schema/types.go]: will fix

File: pkg/mongoproxy/plugins/schema/type_test.go (hunk 2 above)
Comment (yinwang-wish) [reply pkg/mongoproxy/plugins/schema/type_test.go]: since this field is none required, we should not validate it.

File: pkg/mongoproxy/plugins/insort/process.go (hunk 1 above)
Comment (yinwang-wish) [reply pkg/mongoproxy/plugins/insort/process.go]: I saw it on other repo, but can change it to something like : `panic(fmt.Errorf("cannot find correct type in Process:  %s", valTyped))`

File: pkg/mongoproxy/plugins/schema/type_test.go (hunk 2 above)
Comment (zengmin-wish) [reply pkg/mongoproxy/plugins/schema/type_test.go]: required means the field must exist, it doesn't mean it requires validation. All fields in schema will need validation.
